feat(tt-rss) reimplement with upstream container images

ctag opened the pull request:
**Description**
It appears that the k8s-at-home project which this app was based on is unmaintained, and the container has stopped running successfully for myself and others. This pull request moves the tt-rss app back to dev train, and includes an initial attempt at migrating from k8s-at-home to the tt-rss author's docker images.
<!--
Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.
-->

⚒️ Approaches fix for  #8647 <!--(issue)-->

// EDIT: Stavros
Closes #8647

**⚙️ Type of change**

- [ ] ⚙️ Feature/App addition
- [x] 🪛 Bugfix
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 🔃 Refactor of current code

**🧪 How Has This Been Tested?**
Tested with helm+kubernetes on local Docker Desktop instance. Deploys successfully with a local.yaml file which sets LoadBalancer and the Self URL variable:
```
$ cat local.yaml 
service:
  main:
    type: LoadBalancer
workload:
  main:
    podSpec:
      containers:
        main:
          env:
            TTRSS_SELF_URL_PATH: "http://localhost:10104/tt-rss/"
```
<!--
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
-->

**📃 Notes:**
I have not tried implementing updated TrueNAS questions yet, or made updates to the app documentation. The app does not provide semantic versioning that I have found, versions arbitrarily bumped.
<!-- Please enter any other relevant information here -->

**✔️ Checklist:**

- [x] ⚖️ My code follows the style guidelines of this project
- [x] 👀 I have performed a self-review of my own code
- [ ] #️⃣ I have commented my code, particularly in hard-to-understand areas
- [ ] 📄 I have made corresponding changes to the documentation
- [ ] ⚠️ My changes generate no new warnings
- [ ] 🧪 I have added tests to this description that prove my fix is effective or that my feature works
- [ ] ⬆️ I increased versions for any altered app according to semantic versioning

**➕ App addition**

If this PR is an app addition please make sure you have done the following.

- [x] 🪞 I have opened a PR on [truecharts/containers](https://github.com/truecharts/containers) adding the container to TrueCharts mirror repo.
- [x] 🖼️ I have added an icon in the Chart's root directory called `icon.png`

---

_Please don't blindly check all the boxes. Read them and only check those that apply.
Those checkboxes are there for the reviewer to see what is this all about and
the status of this PR with a quick glance._

ctag committed:
moved tt-rss to dev train

ctag committed:
Initial refactor of tt-rss from k8s-at-home to upstream docker images.

Runs on local helm. Still to do:
- Test plugins and theme uploads.
- Implement updated Truenas GUI questions.
- Separate required (self url) and advanced variables.

StevenMcElligott commented:
Move it to Incubator if it's broken and being actively worked on please. Thanks!

ctag committed:
Moved tt-rss to incubator

ctag committed:
Updated app upstream to match ix- prefix.

ctag committed:
Add ADMIN_USER_PASS as exposed configuration option.

ctag committed:
Update questions.yaml with more env variables.

ctag committed:
Merge branch 'master' into ttrss

ctag committed:
Merge branch 'master' into ttrss

StevenMcElligott reviewed (commented):
From the docs part nothing's change, but that's a big rework so you'll have to wait for someone like @stavros-k to take a look, however what's with the extra `local.yaml` file

ctag committed:
Remove local.yaml which was incorrectly included.

ctag committed:
Added initial installation notes and tweaked field name for DNS resolver.

stavros-k commented on charts/incubator/tt-rss/docs/installation_notes.md:
```diff
@@ -0,0 +1,40 @@
+
+tt-rss sets the default nginx resolver to 127.0.0.11, which appears not to work on SCALE apps. The default resolver has been set to TrueNAS SCALE's default of `172.17.0.10`. If a user has changed their SCALE DNS IP, it must be reflected here.
+
+Check the SCALE DNS IP by navigating to `Applications` > `Settings` (dropdown) > `Advanced Settings`. A sidebar titled "Kubernetes Settings" will appear, and toward the bottom is `Cluster DNS IP` with the IP address that should match this app's `TTRSS DNS Resolver` field.
```
Put a HUGE warning that changing the dns in SCALE will **DELETE** their **APPS** and the apps **DATA**. 
Change should only be made in the TTRSS side

Use the docusaurus `:::danger` admonition

ctag replied:
Added a warning in https://github.com/truecharts/charts/pull/17293/commits/f055f6fea9fd2e7307ce292efb5ad25ad458ec9f

Should this section be placed in another file to separate it from common installation instructions?

stavros-k commented on charts/incubator/tt-rss/docs/installation_notes.md:
```diff
@@ -0,0 +1,40 @@
+title: Installation Notes
+---
+
+#### Credentials
```
Whats the reason for `h4`?
Start with `h2` please (it should be the platform target 
And increment from there 

Eg 

`## TrueNAS SCALE`

ctag replied:
Updated headings in https://github.com/truecharts/charts/pull/17293/commits/d865d669f209971e10cb315866d653ae7661bd50

stavros-k commented on charts/incubator/tt-rss/questions.yaml:
```diff
@@ -9,22 +9,48 @@ questions:
+          description: Advanced usage. Set to the default DNS server for TrueNAS Scale.
+          schema:
+            type: string
+            default: "172.17.0.10"
```
Any idea where this resolved is used for?

ctag replied:
My understanding is that this is only used once to set the [resolver field in nginx conf](https://gitlab.tt-rss.org/tt-rss/tt-rss/-/blob/master/.docker/web-nginx/nginx.conf?ref_type=heads#L19).

I expect the default here should actually be "kube-dns.kube-system.svc.cluster.local" to support non-scale use, but I changed it when encountering failures with the probes during this pull request's github checks.

stavros-k commented on charts/incubator/tt-rss/values.yaml:
```diff
@@ -0,0 +1,179 @@
+  #   mountPath: /var/www/html/tt-rss/plugins.local
+  # themes:
+  #   enabled: true
+  #   mountPath: /var/www/html/tt-rss/themes.local
```
What's up with those?

ctag replied:
Left over from the previous implementation. tt-rss now supports installing plugins from a dialog, but I still have a to-do set to figure out how to allow users to manually upload themes to persistent storage.

stavros-k replied:
If you install the plugin from the dialog, where does it gets placed in the filesystem?

ctag replied:
Plugins installed from the dialog are automatically placed in `/var/www/html/tt-rss/plugins.local`. Which means they should be under the html persistent storage.

There's no dialog for themes, which should manually be placed in `/var/www/html/tt-rss/themes.local`. If it's possible to separate this path from the html storage, then users could specify for it to be a separate NFS/etc mount to allow adding themes easily.

stavros-k replied:
You can enable the commented out persistence above, 
should be enough to separate those

stavros-k commented on charts/incubator/tt-rss/values.yaml:
```diff
@@ -0,0 +1,179 @@
+              type: tcp
+              port: "{{ .Values.service.main.ports.main.targetPort }}"
+              # path: "{{ .Values.workload.nginx.podSpec.containers.nginx.env.APP_BASE }}/public.php?op=healthcheck"
+              # port: "{{ .Values.service.main.ports.main.targetPort }}"
```
you dont need the full url.
Only the `/pub...` part

ctag replied:
Thank you. Updated in https://github.com/truecharts/charts/pull/17293/commits/e97e28e4a7571205a964e879592c08932e77915a.

stavros-k commented on charts/incubator/tt-rss/values.yaml:
```diff
@@ -0,0 +1,179 @@
+              port: "{{ .Values.service.app.ports.app.port }}"
+            readiness:
+              type: tcp
+              port: "{{ .Values.service.app.ports.app.port }}"
```
https://gitlab.tt-rss.org/tt-rss/helm-charts/tt-rss/-/blob/master/templates/app-deployment.yaml#L77-83
You can use this for readiness and liveness (assuming is the same image)

ctag replied:
Thank you! Updated in https://github.com/truecharts/charts/pull/17293/commits/a1eda14b489d9045f8136b29ee376398ae039728.

stavros-k commented on charts/incubator/tt-rss/values.yaml:
```diff
@@ -0,0 +1,179 @@
+              port: "{{ .Values.service.main.ports.main.targetPort }}"
+          env:
+            # APP_UPSTREAM: "tt-rss-app.ix-tt-rss.svc.cluster.local"
+            APP_UPSTREAM: '{{ printf "%v-app.ix-%v.svc.cluster.local" (include "tc.v1.common.lib.chart.names.fullname" $) (include "tc.v1.common.lib.chart.names.fullname" $) }}'
```
```suggestion
            APP_UPSTREAM: '{{ printf "%s-app" (include "tc.v1.common.lib.chart.names.fullname" $) }}'
```
This is enough for same namespace resolution.

Also hardcoding `ix-` in namespace is not ok. its not the only platform the charts are used!

ctag committed:
Added danger warning to installation notes that discuss SCALE DNS IP settings.

ctag committed:
Adjusted installation notes headings to begin at H2.

ctag committed:
Changed nginx liveness and readiness checks from tcp to http.

ctag committed:
Update charts/incubator/tt-rss/values.yaml APP_UPSTREAM

Corrects the local name resolution of sibling app to platform independent method.

Co-authored-by: Stavros Kois <47820033+stavros-k@users.noreply.github.com>
Signed-off-by: Christopher <bigbero@gmail.com>

ctag committed:
Merge branch 'master' into ttrss

ctag committed:
Update liveness and readiness checks for app image.

Using: https://gitlab.tt-rss.org/tt-rss/helm-charts/tt-rss/-/blob/master/templates/app-deployment.yaml#L77-83

ctag committed:
Add timeoutSeconds for http liveness check.

ctag committed:
Set APP_UPSTREAM to value recommended by tt-rss author.

https://community.tt-rss.org/t/resolving-issues-with-latest-commit-on-k8s/6208/6

stavros-k commented on charts/incubator/tt-rss/questions.yaml:
```diff
@@ -9,22 +9,48 @@ questions:
+  - variable: env
+    group: App Configuration
+    label: Image Environment
+    schema:
```
The indentation is off! It needs to be a lot more indented (check other apps from stable!)

stavros-k commented on charts/incubator/tt-rss/values.yaml:
```diff
@@ -0,0 +1,181 @@
+              type: tcp
+              port: "{{ .Values.service.main.ports.main.targetPort }}"
+          env:
+            APP_UPSTREAM: '{{ include "tc.v1.common.lib.chart.names.fullname" . }}-app.{{ .Release.Namespace }}.svc.cluster.local'
```
If the it can't resolve a pod in the same namespace without the .NS.svc.cluster.local has other bigger DNS issues.

The "long" format is only for cross namespace, and also the `.svc.cluster.local` is not defacto for all clusters!

ctag replied:
Thank you. I apologize for making you revisit this. 
My current understanding is that nginx's internal resolver does not support search domains (such as in /etc/resolv.conf) and only a FQDN will resolve. I expect that the nginx resolver was added by tt-rss so that a dynamic IP change could be re-resolved because otherwise the value is cached until nginx restarts. 
I've pushed a commit which removes the resolver from nginx.conf, with the expected trade-off that the $fullname-app will be resolved correctly once on startup and then not again.

stavros-k replied:
I don't expect clusterip to change without a recreation of the service or restart of the cluster.
Should be fine!

ctag replied:
Adding links I browsed here, to remove them from the nginx configmap.

nginx resolver requires fqdn: https://stackoverflow.com/a/40695219
variable in proxy_pass forces use of resolver: https://serverfault.com/a/593003
tt-rss discussion: https://community.tt-rss.org/t/resolving-issues-with-latest-commit-on-k8s/6208/7
proxy_pass and resolver: https://stackoverflow.com/a/22259088

stavros-k commented on charts/incubator/tt-rss/values.yaml:
```diff
@@ -0,0 +1,181 @@
+              type: tcp
+              port: "{{ .Values.service.main.ports.main.targetPort }}"
+          env:
+            APP_UPSTREAM: '{{ include "tc.v1.common.lib.chart.names.fullname" . }}-app.{{ .Release.Namespace }}.svc.cluster.local'
```
```suggestion
            APP_UPSTREAM: '{{ printf "%s-app" (include "tc.v1.common.lib.chart.names.fullname" .) }}'
```

stavros-k commented on charts/incubator/tt-rss/values.yaml:
```diff
@@ -0,0 +1,181 @@
+              port: "{{ .Values.service.main.ports.main.targetPort }}"
+          env:
+            APP_UPSTREAM: '{{ include "tc.v1.common.lib.chart.names.fullname" . }}-app.{{ .Release.Namespace }}.svc.cluster.local'
+            RESOLVER: "kube-dns.kube-system.svc.cluster.local"
```
With the app_upstream changed, try to comment out this and see if it can resolve.

ctag committed:
Use configmap to remove nginx resolver.

ctag committed:
Merge in master

ctag committed:
Uncomment persistence for themes and plugins.

ctag committed:
Updated questions indentation to match fireflyiii.

stavros-k commented on charts/incubator/tt-rss/values.yaml:
```diff
@@ -179,3 +187,80 @@ service:
+
+          server {
+            listen 80;
+            listen [::]:80;
```
Lets remove ipv6 and change 80 to `{{ .Values.service.main.ports.main.port }}`

Then remove the `targetPort: 80` completely!

stavros-k commented on charts/incubator/tt-rss/questions.yaml:
```diff
@@ -33,12 +33,6 @@ questions:
-            default: "172.17.0.10"
         - variable: APP_WEB_ROOT
           label: TTRSS App Web Root Path
           description: Advanced usage. This is the filesystem path for tt-rss. Usually set to "/var/www/html", but the default on this app is "/var/www/html/tt-rss". https://tt-rss.org/wiki/InstallationNotes#how-do-i-make-it-run-without-tt-rss-in-the-url-i.e.-at-website-root
```
Lets also remove this and the APP_BASE as it will break persistence if changed

stavros-k commented on charts/incubator/tt-rss/values.yaml:
```diff
@@ -179,3 +187,80 @@ service:
+      nginx.conf.template: |-
+        worker_processes auto;
+        pid /var/run/nginx.pid;
+
```
Lets remove empty lines here
(Mostly because it consumes space when stored in a configmap, and **current** helm versions dont handle the case that you exceed the limit!)

ctag committed:
Update nginx.conf to remove ipv6 and port 80.

ctag committed:
Remove empty lines from configmap.

ctag committed:
Remove APP_WEB_ROOT and APP_BASE from questions.yaml.

Crow-Control reviewed (changes requested):
We NEVER move things to incubator.

It goes from incubator -> stable (or removal)
and from stable -> unstable

Incubator is ONLY for actuall incubation

stavros-k commented:
@ctag Feel free to move it back to stable, as its already marked as breaking change.
(afaik its already broken anyway, right?)

I think we should be good to merge after that.
If you find anything not working as it should we can fix it on another PR!

stavros-k commented on charts/incubator/tt-rss/values.yaml:
```diff
@@ -0,0 +1,242 @@
+updaterImage:
+  repository: cthulhoo/ttrss-fpm-pgsql-static
+  pullPolicy: IfNotPresent
+  tag: 283ad4eb@sha256:f90470ec756dbe8ed9e7f842f19f9c1d7ed4a20aca0a1886e9694b3e7fa82bf7
```
lets change to latest, and let bot update the digest when available.
I don't think it would be able to figure out hashes as tags

ctag committed:
Change images to latest.

ctag committed:
Moved tt-rss to stable.

stavros-k reviewed (approved):
Awesome work!
Thanks

stavros-k committed:
Merge branch 'master' into ttrss

Signed-off-by: Stavros Kois <47820033+stavros-k@users.noreply.github.com>

stavros-k commented:
@all-contributors please add @ctag for code and docs

stavros-k commented on charts/stable/tt-rss/Chart.yaml:
```diff
@@ -5,9 +5,9 @@ annotations:
   truecharts.org/max_helm_version: "3.14"
   truecharts.org/min_helm_version: "3.12"
-  truecharts.org/train: stable
+  truecharts.org/train: incubator
```
dont worry about this, bot will fix it up!

ctag committed:
Merge branch 'master' into ttrss

ctag committed:
Merge branch 'master' into ttrss

ctag committed:
Merge branch 'master' into ttrss

stavros-k committed:
Merge branch 'master' into ttrss

Signed-off-by: Stavros Kois <47820033+stavros-k@users.noreply.github.com>
